x509util/certificate_request.go

```diff
@@ -116,10 +144,123 @@ func (c *CertificateRequest) GetCertificateRequest() (*x509.CertificateRequest,
+		asn1Data, err = c.addChallengePassword(asn1Data)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	// This should not fail
 	return x509.ParseCertificateRequest(asn1Data)
 }
 
+// addChallengePassword unmarshals the asn1Data into a certificateRequest and
+// creates a new one with the challengePassword.
+func (c *CertificateRequest) addChallengePassword(asn1Data []byte) ([]byte, error) {
+	// Marshal challengePassword to ans1.RawValue
+	// Build challengePassword attribute (RFC 2985 section-5.4)
+	var builder cryptobyte.Builder
+	builder.AddASN1(cryptobyte_asn1.SEQUENCE, func(child *cryptobyte.Builder) {
+		child.AddASN1ObjectIdentifier(oidChallengePassword)
+		child.AddASN1(cryptobyte_asn1.SET, func(value *cryptobyte.Builder) {
+			switch {
+			case isPrintableString(c.ChallengePassword, true, true):
+				value.AddASN1(cryptobyte_asn1.PrintableString, func(s *cryptobyte.Builder) {
+					s.AddBytes([]byte(c.ChallengePassword))
+				})
+			case isUTF8String(c.ChallengePassword):
+				value.AddASN1(cryptobyte_asn1.UTF8String, func(s *cryptobyte.Builder) {
+					s.AddBytes([]byte(c.ChallengePassword))
+				})
+			default:
+				value.SetError(errors.New("error marshaling challenge password: password is not valid"))
+			}
+		})
+	})
+
+	b, err := builder.Bytes()
+	if err != nil {
+		return nil, errors.Wrap(err, "error marshaling challenge password")
+	}
+	challengePasswordAttr := asn1.RawValue{
+		FullBytes: b,
+	}
+
+	// Parse certificate request
+	var csr certificateRequest
+
+	rest, err := asn1.Unmarshal(asn1Data, &csr)
+	if err != nil {
+		return nil, err
+	} else if len(rest) != 0 {
+		return nil, errors.New("error unmarshalling certificate request: trailing data")
+	}
+
+	sigAlgo := csr.SignatureAlgorithm
+	tbsCSR := tbsCertificateRequest{
+		Version:       csr.TBSCSR.Version,
+		Subject:       csr.TBSCSR.Subject,
+		PublicKey:     csr.TBSCSR.PublicKey,
+		RawAttributes: csr.TBSCSR.RawAttributes,
+	}
+
+	// Prepend challengePassword attribute
+	tbsCSR.RawAttributes = append([]asn1.RawValue{challengePasswordAttr}, tbsCSR.RawAttributes...)
+
+	// Marshal tbsCertificateRequest
+	tbsCSRContents, err := asn1.Marshal(tbsCSR)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating certificate request")
+	}
+	tbsCSR.Raw = tbsCSRContents
+
+	// Get the hash used previously
+	var hashFunc crypto.Hash
+	found := false
+	sigAlgoOID := sigAlgo.Algorithm
+	for _, m := range signatureAlgorithmMapping {
+		if sigAlgoOID.Equal(m.oid) {
+			hashFunc = m.hash
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, errors.Errorf("error creating certificate request: unsupported signature algorithm %s", sigAlgoOID)
+	}
+
+	// Sign tbsCertificateRequest
+	signed := tbsCSRContents
+	if hashFunc != 0 {
+		h := hashFunc.New()
+		h.Write(signed)
+		signed = h.Sum(nil)
+	}
+
+	var signature []byte
+	signature, err = c.Signer.Sign(rand.Reader, signed, hashFunc)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating certificate request")
```

Comment:
```suggestion
		return nil, fmt.Errorf("error creating certificate request: %", err)
```
Unless you want to keep the stack trace